pkg/controller/kubedirectorcluster/members.go

```diff
@@ -1158,6 +1154,39 @@ func appConfig(
 				}
 				status, convErr := strconv.Atoi(configStatus)
 				if convErr == nil && status == 0 {
+					rs, err := shared.GetRoleStatusByName(cr, roleName)
+					if err != nil {
+						return true, err
+					}
+					upgradingMembers := (*rs).UpgradingMembers
+
+					if upgradingMembers != nil && upgradingMembers[podName] != nil {
```

Comment:
So, here you check for a nil value to decide whether to notify, and also below you set the value to nil when notify has been sent.

Rather than using nil value for this it might be better to use "absence of value" ... so here you would check whether the upgradingMembers[podName] bucket exists in the map, and below you would delete it out of the map. A couple of reasons for that:

- When someone (person or code) is looking at this map to see if a member is currently being upgraded, it's more natural IMO to just see if the member is in the map rather than also having to check the value.
- However, more importantly, I think we can already get into a situation where there are members that aren't in this map, i.e. the role is expanded after the upgrade has been started. So a new member is added to the role (using the new image) and configured, but it is not in this map since it didn't exist when the upgrade was requested. In the current state of the code, that will cause an error on this line when reading upgradingMembers[podName].

So yeah, mostly because of that second reason it seems better to use absence-of-bucket as the representation of "not being upgraded".

Reply:
I'm also thinking about whether there is any case where we can have members left over in this map, that don't get cleared out. That would only be the case for members that are in the process of being set up when the upgrade is triggered. However, that shouldn't be something that can happen since the role upgrade can only start while the cluster is in a stable ready state.

I think that's all good but I will ponder it a bit more.

Reply:
Done